// porg/dist/plugins/auth/controllers/cas/RedirectToCas.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _simpleCasInterface = require('simple-cas-interface');

var _simpleCasInterface2 = _interopRequireDefault(_simpleCasInterface);

var _joi = require('joi');

var _joi2 = _interopRequireDefault(_joi);

var _base64Url = require('base64-url');

var _base64Url2 = _interopRequireDefault(_base64Url);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const handler = casOptions => {
  return (() => {
    var _ref = _asyncToGenerator(function* (request, h) {
      const callback = _base64Url2.default.encode(request.query.callbackUrl);
      const opts = _extends({}, casOptions);
      opts.serviceUrl = opts.serviceUrl + `/${callback}`;

      const cas = new _simpleCasInterface2.default(opts);
      return h.redirect(cas.loginUrl);
    });

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();
};

const config = {
  description: 'Redirects to CAS service',
  validate: {
    query: {
      callbackUrl: _joi2.default.string().required()
    }
  },
  cache: {
    privacy: 'private',
    expiresIn: 0
  },
  plugins: {
    'porg-auth': {
      type: 'no-auth'
    }
  }
};

exports.default = { handler, config };
module.exports = exports['default'];
