// porg/dist/plugins/auth/services/SAMLUtils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function ({ sp, idp }) {
  return {
    sp: new saml2.ServiceProvider({
      entity_id: sp.metadataUrl,
      assert_endpoint: sp.assertUrl,
      private_key: _fs2.default.readFileSync(sp.privateKeyPath).toString(),
      certificate: _fs2.default.readFileSync(sp.certificatePath).toString(),
      allow_unencrypted_assertion: true
    }),
    idp: new saml2.IdentityProvider({
      sso_login_url: idp.loginUrl,
      sso_logout_url: idp.logoutUrl,
      certificates: [_fs2.default.readFileSync(idp.certificatePath).toString()],
      sign_get_request: false,
      allow_unencrypted_assertion: true
    })
  };
};

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _saml2Js = require('saml2-js');

var saml2 = _interopRequireWildcard(_saml2Js);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = exports['default'];
