// porg/src/file-upload-handlers/controllers/CreateUploadSessionController.js
import Joi from 'joi'
import CreateUploadSessionService from '@/file-upload-handlers/services/CreateUploadSessionService'

export default ({ persistenceProvider, fileUploadHandlerConfig }) => {
  const handler = async (request, h) => {
    return CreateUploadSessionService({
      persistenceProvider,
      deduplication: fileUploadHandlerConfig.deduplication,
      checksum: request.payload.checksum
    })
  }

  let maxFileSize = 1048576 // 1MB
  if (fileUploadHandlerConfig && fileUploadHandlerConfig.payload && fileUploadHandlerConfig.payload.maxBytes) {
    maxFileSize = fileUploadHandlerConfig.payload.maxBytes
  }

  const config = {
    description: `Create a new upload session`,
    validate: {
      payload: {
        checksum: Joi.string().required(),
        // Fail fast verification
        fileSize: Joi.number().min(0).max(maxFileSize).required()
      }
    },
    plugins: {
      ...fileUploadHandlerConfig.porgAuth && { 'porg-auth': fileUploadHandlerConfig.porgAuth }
    }
  }

  return {
    handler,
    config
  }
}
