// porg/dist/plugins/auth/controllers/saml/SamlRedirect.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _joi = require('joi');

var _joi2 = _interopRequireDefault(_joi);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const handler = ({ samlOptions }) => {
  return (() => {
    var _ref = _asyncToGenerator(function* (request, h) {
      return new Promise(function (resolve, reject) {
        samlOptions.sp.create_login_request_url(samlOptions.idp, {
          relay_state: request.query.callbackUrl
        }, function (err, loginUrl, requestId) {
          if (err) {
            return reject(err);
          }
          return resolve(h.redirect(loginUrl));
        });
      });
    });

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();
};

const config = {
  description: 'SAML Redirect Handler',
  validate: {
    query: {
      callbackUrl: _joi2.default.string().required()
    }
  },
  cache: {
    privacy: 'private',
    expiresIn: 0
  },
  plugins: {
    'porg-auth': {
      type: 'no-auth'
    }
  }
};

exports.default = { handler, config };
module.exports = exports['default'];
