// porg/dist/plugins/auth/controllers/totp/SubmitTOTPChallengeController.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _joi = require('joi');

var _joi2 = _interopRequireDefault(_joi);

var _otplib = require('otplib');

var _otplib2 = _interopRequireDefault(_otplib);

var _JwtUtils = require('../../services/JwtUtils');

var _errors = require('../../../errors');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const handler = config => {
  return (() => {
    var _ref = _asyncToGenerator(function* (request, h) {
      const totp = yield config.twoFactor.totp.getUserTOTP({ username: request.getPrincipal().username });
      const isValid = _otplib2.default.authenticator.verify({ secret: totp.secret, token: request.payload.code });
      if (isValid) {
        const token = yield (0, _JwtUtils.getJwt)({
          payload: _extends({}, request.getPrincipal(), { twoFactor: true }),
          secretOrPrivateKey: config.jwt.secret,
          options: { expiresIn: config.jwt.expires || '1 hour' }
        });
        h.state('P_SESSION', token, { encoding: 'none', ttl: config.sessionTime * 1000, isSameSite: 'Strict', path: '/' });
        h.state('P_INFO', Buffer.from(JSON.stringify({ twoFactor: false, roles: request.getPrincipal().roles })).toString('base64'), { encoding: 'none', ttl: config.sessionTime * 1000, isSameSite: 'Strict', path: '/', isHttpOnly: false });
        return {}; // returning h.state was breaking the endpoint
      } else {
        throw (0, _errors.errorWithKey)('not-authorized', {
          ctx: {}
        });
      }
    });

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();
};

const config = {
  description: 'Submits a TOTP challenge code to verify',
  validate: {
    payload: {
      code: _joi2.default.number().integer().required()
    }
  },
  plugins: {
    'porg-auth': {
      type: 'user-session',
      roles: []
    }
  }
};

exports.default = { handler, config };
module.exports = exports['default'];
