// porg/dist/providers/storage/FenixEduDriveStorageProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

class FenixEduDriveStorageProvider {
  constructor({ rootPath }) {
    this.rootPath = rootPath;
  }

  uploadFile({ file, path }) {
    return _asyncToGenerator(function* () {})();
  }

  downloadFile({ filePath }) {
    return _asyncToGenerator(function* () {})();
  }

  deleteFile({ filePath }) {
    return _asyncToGenerator(function* () {})();
  }

  listFiles({ searchPath }) {
    return _asyncToGenerator(function* () {})();
  }
}
exports.default = FenixEduDriveStorageProvider;
module.exports = exports["default"];
