// porg/dist/plugins/healthcheck/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.plugin = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _StatusHealthcheckController = require('./controllers/StatusHealthcheckController');

var _StatusHealthcheckController2 = _interopRequireDefault(_StatusHealthcheckController);

var _StatsHealthcheckController = require('./controllers/StatsHealthcheckController');

var _StatsHealthcheckController2 = _interopRequireDefault(_StatsHealthcheckController);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const plugin = exports.plugin = {
  name: 'healthcheck-plugin',
  version: '1.0.0',
  register: function (server, options) {
    server.route(_extends({
      path: '/api/v1/healthcheck/status',
      method: 'GET'
    }, (0, _StatusHealthcheckController2.default)(options)));

    server.route(_extends({
      path: '/api/v1/healthcheck/stats',
      method: 'GET'
    }, (0, _StatsHealthcheckController2.default)(options)));
  }
};
