// porg/dist/tasks/controllers/GetInstanceTaskById.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _joi = require('joi');

var _joi2 = _interopRequireDefault(_joi);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const handler = ({ persistentTaskProvider }) => {
  return (() => {
    var _ref = _asyncToGenerator(function* (request, h) {
      const task = yield persistentTaskProvider.getTaskInstanceById({ id: request.params.id });
      return task;
    });

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();
};

const config = ({ roles, twoFactor }) => {
  return {
    description: 'Get task instance by id',
    validate: {
      params: {
        id: _joi2.default.string().required()
      }
    },
    plugins: {
      'porg-auth': {
        type: 'user-session',
        twoFactor,
        roles
      }
    }
  };
};

exports.default = { handler, config };
module.exports = exports['default'];
