// porg/dist/tasks/controllers/GetInstanceTasks.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _StatusTask = require('../StatusTask');

var _StatusTask2 = _interopRequireDefault(_StatusTask);

var _joi = require('joi');

var _joi2 = _interopRequireDefault(_joi);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const handler = ({ persistentTaskProvider }) => {
  return (() => {
    var _ref = _asyncToGenerator(function* (request, h) {
      const tasks = yield persistentTaskProvider.getTasksInstances({ status: request.query.status });
      return { tasks: tasks };
    });

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();
};

const config = ({ roles, twoFactor }) => {
  return {
    description: 'Get tasks instances',
    validate: {
      query: {
        status: _joi2.default.string().valid(Object.values(_StatusTask2.default)).required()
      }
    },
    plugins: {
      'porg-auth': {
        type: 'user-session',
        twoFactor,
        roles
      }
    }
  };
};

exports.default = { handler, config };
module.exports = exports['default'];
