// porg/dist/plugins/auth/controllers/u2f/GetU2FRegistrationChallenge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _u2f = require('u2f');

var _u2f2 = _interopRequireDefault(_u2f);

var _UserHasAnyTwoFactorAuth = require('../../services/UserHasAnyTwoFactorAuth');

var _UserHasAnyTwoFactorAuth2 = _interopRequireDefault(_UserHasAnyTwoFactorAuth);

var _errors = require('../../../errors');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const handler = config => {
  return (() => {
    var _ref = _asyncToGenerator(function* (request, h) {
      const username = request.getPrincipal().username;
      const userAlreadyHasTwoFactor = yield (0, _UserHasAnyTwoFactorAuth2.default)({
        username,
        getUserU2F: config.twoFactor.u2f.getUserU2F,
        getUserTOTP: config.twoFactor.totp.getUserTOTP
      });

      if (userAlreadyHasTwoFactor && !request.auth.credentials.twoFactor) {
        throw (0, _errors.errorWithKey)('requires-two-factor-challenge', {
          ctx: {}
        });
      }
      let authRequest = _u2f2.default.request(config.twoFactor.u2f.appId);
      yield config.twoFactor.u2f.upsertUserU2F({ username, u2f: { authRequest } });
      return authRequest;
    });

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();
};

const config = {
  description: 'Get registration challenge for U2F',
  plugins: {
    'porg-auth': {
      type: 'user-session',
      roles: []
    }
  }
};

exports.default = { handler, config };
module.exports = exports['default'];
