// porg/dist/plugins/auth/controllers/u2f/SubmitU2FRegistrationChallenge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _u2f = require('u2f');

var _u2f2 = _interopRequireDefault(_u2f);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

// import { errorWithKey } from '@/plugins/erros'

const handler = config => {
  return (() => {
    var _ref = _asyncToGenerator(function* (request, h) {
      const credentials = request.getPrincipal();
      let userU2F = yield config.twoFactor.u2f.getUserU2F({ username: credentials.username });
      var checkRes = _u2f2.default.checkRegistration(userU2F.authRequest, request.payload);
      if (checkRes.successful) {
        yield config.twoFactor.u2f.saveUserU2FChallenge({
          username: credentials.username,
          u2f: {
            publicKey: checkRes.publicKey,
            keyHandle: checkRes.keyHandle,
            certificate: checkRes.certificate
          }
        });
        return true;
      } else {
        return false;
      }
    });

    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();
};

const config = {
  description: 'Submit registration challenge for U2F',
  plugins: {
    'porg-auth': {
      type: 'user-session',
      roles: []
    }
  }
};

exports.default = { handler, config };
module.exports = exports['default'];
