// metadata-validator/dist/parsers/InProceedingsParser.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _ParserGenerator = require('./ParserGenerator');

var _ParserGenerator2 = _interopRequireDefault(_ParserGenerator);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = validators => {
  return (0, _ParserGenerator2.default)({
    title: {
      required: true,
      type: String,
      validator: validators.title
    },
    year: {
      required: true,
      type: Number,
      validator: validators.year
    },
    event: {
      required: true,
      type: String,
      validator: validators.event
    },
    author: {
      required: true,
      type: Object,
      validator: validators.authorWithAtLeastOneLink
    },
    abstract: {
      required: false,
      type: String,
      validator: validators.abstract
    },
    doi: {
      required: false,
      type: String,
      validator: validators.doi
    },
    series: {
      required: false,
      type: String,
      validator: validators.series
    },
    pages: {
      required: false,
      type: String,
      validator: validators.pages
    },
    address: {
      required: false,
      type: String,
      validator: validators.address
    },
    month: {
      required: false,
      type: Number,
      validator: validators.month
    },
    organization: {
      required: false,
      type: String,
      validator: validators.organization
    },
    publisher: {
      required: false,
      type: String,
      validator: validators.publisher
    },
    url: {
      required: false,
      type: String,
      validator: validators.url
    },
    keyword: {
      required: false,
      type: Array,
      validator: validators.keyword
    },
    template: {
      required: false,
      type: String,
      validator: validators.template
    }
  });
};

module.exports = exports['default'];
