// porg/dist/drivers/Smtp.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.shutdown = exports.transporter = exports.cleanDatabase = exports.configDriver = undefined;

var _nodemailer = require('nodemailer');

var _nodemailer2 = _interopRequireDefault(_nodemailer);

var _papagaio = require('papagaio');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

let transporter;
const configDriver = ({ options }) => {
  try {
    exports.transporter = transporter = _nodemailer2.default.createTransport({
      host: options.host,
      port: options.port,
      secure: options.secure
    });
  } catch (err) {
    (0, _papagaio.logger)(['porg', 'err', 'email'], `Issues configurating email `);
    throw new Error(err);
  }
};

const cleanDatabase = (() => {
  var _ref = _asyncToGenerator(function* () {
    // notting to do
  });

  return function cleanDatabase() {
    return _ref.apply(this, arguments);
  };
})();

const shutdown = (() => {
  var _ref2 = _asyncToGenerator(function* () {});

  return function shutdown() {
    return _ref2.apply(this, arguments);
  };
})();

exports.configDriver = configDriver;
exports.cleanDatabase = cleanDatabase;
exports.transporter = transporter;
exports.shutdown = shutdown;
