// porg/dist/file-upload-handlers/controllers/UploadFileController.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _joi = require('joi');

var _joi2 = _interopRequireDefault(_joi);

var _UploadFileService = require('../services/UploadFileService');

var _UploadFileService2 = _interopRequireDefault(_UploadFileService);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

exports.default = ({ persistenceProvider, StorageProvider, fileUploadHandlerConfig }) => {
  const handler = (() => {
    var _ref = _asyncToGenerator(function* (request, h) {
      return (0, _UploadFileService2.default)({
        persistenceProvider,
        StorageProvider,
        tempStorageRootPath: fileUploadHandlerConfig.tempStorageRootPath,
        mimeTypes: fileUploadHandlerConfig.mimeTypes,
        deduplication: fileUploadHandlerConfig.deduplication,
        session: request.params.id,
        file: request.payload.file
      });
    });

    return function handler(_x, _x2) {
      return _ref.apply(this, arguments);
    };
  })();

  const config = {
    description: 'Uploads a file to an upload session',
    validate: {
      params: {
        id: _joi2.default.string().required()
      }
    },
    plugins: _extends({}, fileUploadHandlerConfig.porgAuth && { 'porg-auth': fileUploadHandlerConfig.porgAuth }),
    payload: _extends({
      output: 'stream',
      parse: true,
      allow: 'multipart/form-data'
    }, fileUploadHandlerConfig.payload && _extends({}, fileUploadHandlerConfig.payload))
  };

  return {
    handler,
    config
  };
};

module.exports = exports['default'];
