// porg/src/drivers/index.js
import * as Mongo from '@/drivers/MongoDB'
import * as Elasticsearch from '@/drivers/Elasticsearch'
import * as Consul from '@/drivers/Consul'
import * as Neo4J from '@/drivers/Neo4J'
import * as MySQL from '@/drivers/MySQL'
import * as Smtp from '@/drivers/Smtp'
import * as S3 from '@/drivers/S3'
import * as FenixEduDrive from '@/drivers/FenixEduDrive'
import { errorWithKey } from '@/plugins/errors'

const driversMap = {
  mongo: Mongo,
  elasticsearch: Elasticsearch,
  consul: Consul,
  mysql: MySQL,
  neo4j: Neo4J,
  smtp: Smtp,
  s3: S3,
  fenixeduDrive: FenixEduDrive
}

const runConfigWithDriver = ({ driver, options }) => {
  const driverEntry = driversMap[driver]
  if (driverEntry) {
    driverEntry.configDriver({ options: options })
    return driverEntry
  } else {
    return errorWithKey('driver-not-found', {
      tags: ['error', 'database', 'driver'],
      ctx: { driver },
      metadata: { driver }
    })
  }
}

export {
  runConfigWithDriver,
  Mongo,
  Elasticsearch,
  Neo4J,
  Consul,
  MySQL,
  Smtp
}
