// porg/test/TestTemplate.js
import test from 'tape-promise/tape'
import { config, logger, setConfig, porg, TestUtils, consoleFormat } from '@/'
process.on('unhandledRejection', (reason, p) => {
  // application specific logging, throwing an error, or other logic here
  console.log(reason)
  logger(['error', 'testing', 'migration'], `Unhandled Rejection at: Promise ${p} reason ${JSON.stringify(reason)}`)
})

let server
setConfig({basePath: `${__dirname}/env/normal`, redefine: true})
const testCaseRunner = (name, env, initRoutes, cb) => {
  if (!cb) {
    cb = initRoutes
    initRoutes = () => {}
  }
  test(name, function (t) {
    t.test('setup', async function (t) {
      setConfig({basePath: `${__dirname}/env/${env}`, redefine: true})
      logger(['info', 'testing', 'test'], `Setup testing`)
      await TestUtils.cleanDbs({config})
      const porgServer = await porg({
        runOnce: () => {
        }
      })
      server = porgServer.getHapiServer()
      initRoutes(server)
      await porgServer.start()
      logger(['info', 'testing', 'test'], `Server finish init`)
      t.end()
    })

    cb(t)

    t.test('teardown', async function (t) {
      logger(['info', 'testing', 'test'], `Teardown testing`)
      await server.stop()
      await TestUtils.closeDbs({config})
      logger(['info', 'testing', 'test'], `Finish testing`)
      t.end()
    })
  })
}

export {
  logger,
  consoleFormat,
  testCaseRunner,
  server
}
